src/bson/ext.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright 2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use mongodb::bson::{document::ValueAccessError, from_bson, from_document, Bson, Document};
+use serde::de::DeserializeOwned;
+use thiserror::Error;
+
+/// Gets values and upcasts if necessary
+#[allow(missing_docs)]
+pub trait BsonExt {
+    fn as_string(&self) -> Result<String, ValueAccessError>;
+
+    fn as_u8(&self) -> Result<u8, ValueAccessError>;
+
+    fn as_u16(&self) -> Result<u16, ValueAccessError>;
+
+    fn as_u32(&self) -> Result<u32, ValueAccessError>;
+
+    fn as_u64(&self) -> Result<u64, ValueAccessError>;
+
+    fn to_bytes(self) -> Result<Vec<u8>, ValueAccessError>;
+
+    fn to_array(self) -> Result<Vec<Bson>, ValueAccessError>;
+
+    fn to_document(self) -> Result<Document, ValueAccessError>;
+}
+
+impl BsonExt for Bson {
+    fn as_string(&self) -> Result<String, ValueAccessError> {
+        Ok(match self {
+            Bson::Double(i) => i.to_string(),
+            Bson::String(i) => i.to_string(),
+            Bson::Document(i) => i.to_string(),
+            Bson::Boolean(i) => i.to_string(),
+            Bson::Null => "null".to_string(),
+            Bson::RegularExpression(i) => i.to_string(),
+            Bson::JavaScriptCode(i) => i.to_string(),
+            Bson::JavaScriptCodeWithScope(i) => i.to_string(),
+            Bson::Int32(i) => i.to_string(),
+            Bson::Int64(i) => i.to_string(),
+            Bson::Timestamp(i) => i.to_string(),
+            Bson::Binary(i) => i.to_string(),
+            Bson::ObjectId(i) => i.to_string(),
+            Bson::DateTime(i) => i.to_string(),
+            Bson::Symbol(i) => i.to_string(),
+            Bson::Decimal128(i) => i.to_string(),
+            Bson::Undefined => "undefined".to_string(),
+            _ => return Err(ValueAccessError::UnexpectedType),
+        })
+    }
+
+    fn as_u8(&self) -> Result<u8, ValueAccessError> {
+        Ok(match self {
+            Bson::Double(i) => *i as u8,
+            Bson::Boolean(i) => *i as u8,
+            Bson::Int32(i) => *i as u8,
+            Bson::Int64(i) => *i as u8,
```

Comment:
I guess the circumstances in which those extension methods are used already ensure correct usage, i.e. that those won't panic? Or should we do manual bound checks and return an `ValueAccessError::OutOfBounds` or something to ensure the conversion is always possible? Just picked `as_u8` as an example here.

Reply:
I think we absolutely could, but for the purpose of this helper trait it is unnecessary. The intention for this trait is to use it for values that are known.